x/expiration/types/expiration.go

```diff
@@ -0,0 +1,100 @@
+package types
+
+import (
+	"fmt"
+	"regexp"
+	"strconv"
+	"strings"
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/codec/types"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+)
+
+func NewExpiration(
+	moduleAssetID string,
+	owner string,
+	time time.Time,
+	deposit sdk.Coins,
+	message types.Any,
+) *Expiration {
+	return &Expiration{
+		ModuleAssetId: moduleAssetID,
+		Owner:         owner,
+		Time:          time,
+		Deposit:       deposit,
+		Message:       message,
+	}
+}
+
+// ValidateBasic basic format checking of the data
+func (e *Expiration) ValidateBasic() error {
+	if strings.TrimSpace(e.ModuleAssetId) == "" {
+		return ErrEmptyModuleAssetID
+	}
+	if err := validateAddress(e.ModuleAssetId); err != nil {
+		return err
+	}
+	if strings.TrimSpace(e.Owner) == "" {
+		return ErrEmptyOwnerAddress
+	}
+	if e.Time.IsZero() {
+		return ErrExpirationTime
+	}
+	if e.Deposit.IsZero() || e.Deposit.IsAnyNil() || !e.Deposit.IsValid() {
+		return ErrInvalidDeposit
+	}
+	if err := e.validateMessage(); err != nil {
+		return ErrInvalidMessage
+	}
+	return nil
+}
+
+// validateMessage validates `expiration.Message` conforms to `sdk.Msg`
+// and is registered and whitelisted with the InterfaceRegistry
+func (e *Expiration) validateMessage() error {
+	// this may occur in message decoding
+	if e.Message.Value == nil {
+		return fmt.Errorf("expecting non nil value to validate an Any message")
+	}
+	if e.Message.TypeUrl == "" {
+		return fmt.Errorf("expecting message type URL to unpack Any")
+	}
+	return nil
```

Comment:
Expiration looks to contain a `sdk.Msg` that will get dispatched later.  We should probably call the `ValidateBasic()` on the msg as well.

Reply:
I agree.... the nested msgs should have their ValidateBasic called by the parent msg ValidateBasic flow... beyond this check no deeper validation on the nested msg would be appropriate.

Reply:
Yes this one is a bit tricky to solve.

I initially was registering the `delete` Msg types with the codec to be able to Unmashal and call `ValidateBasic()` on the `msg` but we don't want the expiration module to have references to other modules as it may lead to circular dependencies (ref: https://github.com/provenance-io/provenance/pull/1115#discussion_r987141369). I have an open ticket to address this one based on what @SpicyLemon and I discussed.


```go
func whiteListedInterfaceRegistry() types.InterfaceRegistry {
 	ir := types.NewInterfaceRegistry()
 	ir.RegisterImplementations((*sdk.Msg)(nil),
 		&metadatatypes.MsgDeleteScopeRequest{},
 		&metadatatypes.MsgDeleteRecordRequest{},
...
```

https://github.com/provenance-io/provenance/blob/9f743b422abde49db364c95211ecd9524d9a1ff6/x/expiration/client/cli/tx.go#L287-L293

Reply:
Actually it's simpler than I thought. We can call `msg.ValidateBasic()` in `SetExpiration`. Same as we do in `InvokeExpiration`.

Reply:
added `ValidateBasic()` in the keeper. 

https://github.com/provenance-io/provenance/blob/19afc634a6204f27598350ce59ecd578419ee2bd/x/expiration/keeper/keeper.go#L136-L149